### PR TITLE
Build: Try to lazy load moment locales

### PR DESCRIPTION
**Pull request by blowery**
At the moment, we bundle all of the moment locales into the calypso vendor chunk. We only really need one or two of them most of the time, so instead, lazy load only what we need.

## Complications
Before this patch, all the locales are bundled with moment, so when we switch locales, we already have the bundles loaded. After this patch, there's a window where we've set the locale, but we have _not_ loaded the moment locale yet. That could result in weird formatting and possibly broken date parsing if we're parsing strings that depend on the current locale being correct. This is similar to how translations currently load.

Unlike translations, we currently have no way of telling everything using moment that it needs to re-render. I'd like to work on adding some reactive date formatting controls in a future PR.

## Testing
English should be just as it is now, as those translations are still bundled.
Switch your locale to something else and reload Calypso. Verify that dates being provided by moment are using the new locale. A good place to check is the Reader's date stamps on each card.

### CODE REVIEW

**Comment by flootr** [comment]
I could spot the following when I switch my account language to "German":

- **Reader**: it should say `8. Nov. 2017` but what it shows is `Nov 8, 2017`. It works, however, for the string `3 days ago`, there it says `Vor 3 Tag(en)` which sounds correct.
- **Account Settings**: it should say `Beigetreten im Mai 2014` but it shows `Beigetreten im May 2014`.
- **Stats**: Timeline dates are not formatted correctly, e.g. it says `7. Jul` but correct is `7. Juli`.

Note: I compared calypso.live with prod and did a reload before checking.

**Comment by blowery** [comment]
@flootr repro'd. hrm.... digging into it.

**Comment by blowery** [comment]
Looks like our `asyncRequire` doesn't work with really dynamic includes. This is a bit of [special webpack magic](https://webpack.js.org/api/module-methods/#import-).

**Comment by blowery** [comment]
New stats
```
Delta:
chunk                                                                                                          stat_size            parsed_size            gzip_size
async-load-design-blocks                                                                                            +0 B                   +0 B                 +1 B   (+0.0%)
build                                                                                                            -2169 B   (-0.1%)      +2036 B   (+0.1%)     +502 B   (+0.2%)
devdocs                                                                                                             +0 B                   +0 B                 +2 B   (+0.0%)
manifest                                                                                                            +0 B                +3321 B  (+23.8%)    +1933 B  (+34.1%)
posts-custom~posts-pages                                                                                            +0 B                   +0 B                 +1 B   (+0.0%)
vendors~async-load-blocks-calendar-popover~async-load-design~async-load-design-blocks~async-load-des~7fc5def2    +2958 B   (+1.4%)      +1300 B   (+1.2%)     +581 B   (+3.1%)
vendors~build                                                                                                  -391246 B  (-13.5%)    -182896 B  (-13.9%)   -48811 B  (-14.8%)
171                                                                                                              +2577 B     (new)      +1314 B     (new)     +722 B     (new)
172                                                                                                              +2004 B     (new)      +1111 B     (new)     +681 B     (new)
173                                                                                                              +1995 B     (new)      +1102 B     (new)     +677 B     (new)
174                                                                                                              +4043 B     (new)      +2013 B     (new)    +1104 B     (new)
175                                                                                                              +1997 B     (new)      +1102 B     (new)     +677 B     (new)
176                                                                                                              +3089 B     (new)      +1551 B     (new)     +914 B     (new)
177                                                                                                              +1965 B     (new)      +1103 B     (new)     +675 B     (new)
178                                                                                                              +4336 B     (new)      +2141 B     (new)    +1178 B     (new)
179                                                                                                              +3311 B     (new)      +1659 B     (new)     +914 B     (new)
180                                                                                                              +4849 B     (new)      +2256 B     (new)    +1293 B     (new)
181                                                                                                              +3101 B     (new)      +1380 B     (new)     +910 B     (new)
182                                                                                                              +2079 B     (new)      +1194 B     (new)     +630 B     (new)
183                                                                                                              +3567 B     (new)      +1634 B     (new)    +1042 B     (new)
184                                                                                                              +3818 B     (new)      +1885 B     (new)    +1066 B     (new)
185                                                                                                              +3427 B     (new)      +1534 B     (new)     +833 B     (new)
186                                                                                                              +5216 B     (new)      +1963 B     (new)     +877 B     (new)
187                                                                                                              +3300 B     (new)      +1799 B     (new)     +853 B     (new)
188                                                                                                              +6997 B     (new)      +2431 B     (new)    +1061 B     (new)
189                                                                                                              +2288 B     (new)      +1298 B     (new)     +814 B     (new)
190                                                                                                              +2911 B     (new)      +1405 B     (new)     +757 B     (new)
191                                                                                                              +2041 B     (new)      +1135 B     (new)     +628 B     (new)
192                                                                                                              +2818 B     (new)      +1447 B     (new)     +754 B     (new)
193                                                                                                              +2811 B     (new)      +1440 B     (new)     +740 B     (new)
194                                                                                                              +2811 B     (new)      +1441 B     (new)     +748 B     (new)
195                                                                                                              +2646 B     (new)      +1243 B     (new)     +814 B     (new)
196                                                                                                              +3809 B     (new)      +1980 B     (new)    +1215 B     (new)
197                                                                                                              +2323 B     (new)      +1238 B     (new)     +682 B     (new)
198                                                                                                              +2151 B     (new)      +1222 B     (new)     +673 B     (new)
199                                                                                                              +2319 B     (new)      +1234 B     (new)     +682 B     (new)
200                                                                                                              +2318 B     (new)      +1233 B     (new)     +682 B     (new)
201                                                                                                              +2113 B     (new)      +1199 B     (new)     +663 B     (new)
202                                                                                                              +2323 B     (new)      +1238 B     (new)     +682 B     (new)
203                                                                                                              +2479 B     (new)      +1310 B     (new)     +730 B     (new)
204                                                                                                              +3749 B     (new)      +2177 B     (new)     +940 B     (new)
205                                                                                                              +3017 B     (new)      +1606 B     (new)     +772 B     (new)
206                                                                                                              +3741 B     (new)      +2163 B     (new)     +936 B     (new)
207                                                                                                              +3110 B     (new)      +1455 B     (new)     +763 B     (new)
208                                                                                                              +2352 B     (new)      +1360 B     (new)     +688 B     (new)
209                                                                                                              +3225 B     (new)      +1633 B     (new)     +931 B     (new)
210                                                                                                              +4010 B     (new)      +1944 B     (new)     +890 B     (new)
211                                                                                                              +2062 B     (new)      +1156 B     (new)     +652 B     (new)
212                                                                                                              +2525 B     (new)      +1321 B     (new)     +708 B     (new)
213                                                                                                              +2700 B     (new)      +1340 B     (new)     +719 B     (new)
214                                                                                                              +2958 B     (new)      +1358 B     (new)     +721 B     (new)
215                                                                                                              +2669 B     (new)      +1391 B     (new)     +764 B     (new)
216                                                                                                              +2575 B     (new)      +1346 B     (new)     +757 B     (new)
217                                                                                                              +2759 B     (new)      +1481 B     (new)     +734 B     (new)
218                                                                                                              +4516 B     (new)      +1944 B     (new)     +958 B     (new)
219                                                                                                              +3932 B     (new)      +1680 B     (new)    +1080 B     (new)
220                                                                                                              +3258 B     (new)      +1522 B     (new)     +872 B     (new)
221                                                                                                              +3830 B     (new)      +1636 B     (new)    +1067 B     (new)
222                                                                                                              +5396 B     (new)      +2106 B     (new)     +939 B     (new)
223                                                                                                              +4142 B     (new)      +1945 B     (new)     +926 B     (new)
224                                                                                                              +3213 B     (new)      +1624 B     (new)     +960 B     (new)
225                                                                                                              +2779 B     (new)      +1333 B     (new)     +712 B     (new)
226                                                                                                              +4804 B     (new)      +1828 B     (new)     +891 B     (new)
227                                                                                                              +2370 B     (new)      +1279 B     (new)     +713 B     (new)
228                                                                                                              +2715 B     (new)      +1253 B     (new)     +722 B     (new)
229                                                                                                              +2844 B     (new)      +1398 B     (new)     +733 B     (new)
230                                                                                                              +3090 B     (new)      +1683 B     (new)     +990 B     (new)
231                                                                                                              +2635 B     (new)      +1390 B     (new)     +882 B     (new)
232                                                                                                              +3131 B     (new)      +1565 B     (new)    +1003 B     (new)
233                                                                                                              +3882 B     (new)      +1800 B     (new)    +1142 B     (new)
234                                                                                                              +2453 B     (new)      +1232 B     (new)     +728 B     (new)
235                                                                                                              +2642 B     (new)      +1397 B     (new)     +890 B     (new)
236                                                                                                              +4827 B     (new)      +1804 B     (new)     +943 B     (new)
237                                                                                                              +2292 B     (new)      +1295 B     (new)     +810 B     (new)
238                                                                                                              +4452 B     (new)      +2242 B     (new)    +1097 B     (new)
239                                                                                                              +3891 B     (new)      +1830 B     (new)     +844 B     (new)
240                                                                                                              +4140 B     (new)      +2016 B     (new)     +950 B     (new)
241                                                                                                              +2348 B     (new)      +1437 B     (new)     +743 B     (new)
242                                                                                                              +3113 B     (new)      +1392 B     (new)     +905 B     (new)
243                                                                                                              +2768 B     (new)      +1438 B     (new)     +955 B     (new)
244                                                                                                              +3534 B     (new)      +1737 B     (new)     +965 B     (new)
245                                                                                                              +5404 B     (new)      +2360 B     (new)    +1268 B     (new)
246                                                                                                              +2798 B     (new)      +1352 B     (new)     +717 B     (new)
247                                                                                                              +2791 B     (new)      +1346 B     (new)     +714 B     (new)
248                                                                                                              +2038 B     (new)      +1132 B     (new)     +675 B     (new)
249                                                                                                              +2701 B     (new)      +1405 B     (new)     +928 B     (new)
250                                                                                                              +2123 B     (new)      +1197 B     (new)     +640 B     (new)
251                                                                                                              +3770 B     (new)      +1732 B     (new)    +1100 B     (new)
252                                                                                                              +3424 B     (new)      +1978 B     (new)     +943 B     (new)
253                                                                                                              +3417 B     (new)      +1972 B     (new)     +941 B     (new)
254                                                                                                              +2058 B     (new)      +1158 B     (new)     +635 B     (new)
255                                                                                                              +3950 B     (new)      +1651 B     (new)    +1015 B     (new)
256                                                                                                              +4628 B     (new)      +2014 B     (new)     +990 B     (new)
257                                                                                                              +2154 B     (new)      +1271 B     (new)     +694 B     (new)
258                                                                                                              +2305 B     (new)      +1267 B     (new)     +696 B     (new)
259                                                                                                              +2552 B     (new)      +1195 B     (new)     +708 B     (new)
260                                                                                                              +7447 B     (new)      +3444 B     (new)    +1608 B     (new)
261                                                                                                              +2508 B     (new)      +1133 B     (new)     +750 B     (new)
262                                                                                                              +2174 B     (new)      +1275 B     (new)     +703 B     (new)
263                                                                                                              +2378 B     (new)      +1329 B     (new)     +860 B     (new)
264                                                                                                              +5997 B     (new)      +2095 B     (new)     +927 B     (new)
265                                                                                                              +7002 B     (new)      +2448 B     (new)     +992 B     (new)
266                                                                                                              +2284 B     (new)      +1245 B     (new)     +697 B     (new)
267                                                                                                              +4128 B     (new)      +2003 B     (new)    +1086 B     (new)
268                                                                                                              +4120 B     (new)      +1998 B     (new)     +947 B     (new)
269                                                                                                              +3056 B     (new)      +1502 B     (new)     +799 B     (new)
270                                                                                                              +2368 B     (new)      +1250 B     (new)     +695 B     (new)
271                                                                                                              +2042 B     (new)      +1144 B     (new)     +626 B     (new)
272                                                                                                              +4165 B     (new)      +1885 B     (new)    +1157 B     (new)
273                                                                                                              +3050 B     (new)      +1454 B     (new)     +960 B     (new)
274                                                                                                              +2325 B     (new)      +1246 B     (new)     +684 B     (new)
275                                                                                                              +3564 B     (new)      +1641 B     (new)     +985 B     (new)
276                                                                                                              +2291 B     (new)      +1306 B     (new)     +857 B     (new)
277                                                                                                              +2134 B     (new)      +1198 B     (new)     +652 B     (new)
278                                                                                                              +4157 B     (new)      +2055 B     (new)     +889 B     (new)
279                                                                                                              +2958 B     (new)      +1441 B     (new)     +799 B     (new)
280                                                                                                              +3585 B     (new)      +1582 B     (new)     +856 B     (new)
281                                                                                                              +2021 B     (new)      +1136 B     (new)     +560 B     (new)
282                                                                                                              +1992 B     (new)      +1111 B     (new)     +647 B     (new)
283                                                                                                              +3907 B     (new)      +1804 B     (new)    +1015 B     (new)
284                                                                                                              +5562 B     (new)      +2479 B     (new)    +1420 B     (new)
285                                                                                                              +2509 B     (new)      +1134 B     (new)     +726 B     (new)
286                                                                                                              +2024 B     (new)      +1135 B     (new)     +595 B     (new)
287                                                                                                              +1985 B     (new)      +1100 B     (new)     +706 B     (new)
288                                                                                                              +2672 B     (new)      +1433 B     (new)     +764 B     (new)
289                                                                                                              +2456 B     (new)      +1357 B     (new)     +850 B     (new)
290                                                                                                              +2101 B     (new)      +1195 B     (new)     +700 B     (new)
291                                                                                                              +3467 B     (new)      +1397 B     (new)     +915 B     (new)
292                                                                                                              +3209 B     (new)      +1394 B     (new)     +901 B     (new)
293                                                                                                              +3214 B     (new)      +1399 B     (new)     +902 B     (new)
```

**Comment by blowery** [comment]
@sirreal looks like we might need to bump the version of babel that the pot generator uses? https://circleci.com/gh/Automattic/wp-calypso/99926?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by blowery** [comment]
@sirreal ah, it's part of i18n-calypso@2! trying a bump to see if it fixes it.

**Comment by blowery** [comment]
@sirreal a new and exciting error
`SyntaxError: This experimental syntax requires enabling the parser plugin: 'exportDefaultFrom' (2:7)`

**Comment by blowery** [comment]
https://github.com/Automattic/i18n-calypso/pull/68 should fix the parser

**Comment by sirreal** [comment]
What a rollercoaster!

https://github.com/Automattic/i18n-calypso/pull/68 fixes the issue you uncovered, I pushed a temporary commit to test with the git hash (4fe32ed)

You can see that the previously failing step now passes. The following step (*Build New Strings .pot*) errors but does not report failure 😕 

https://circleci.com/gh/Automattic/wp-calypso/99977

The issues with the following step would likely be fixed by https://github.com/Automattic/gp-localci-client/pull/12

**Comment by sirreal** [comment]
Everything looks good with the two temporary testing commits:

https://circleci.com/gh/Automattic/wp-calypso/99980

We'd need to:
- [ ] Release and bump to `i18n-calypso@2.0.1` with https://github.com/Automattic/i18n-calypso/pull/68
- [ ] Review and merge https://github.com/Automattic/gp-localci-client/pull/12

**Comment by sirreal** [comment]
Here's a fix for the error that doesn't cause the job to fail:

https://github.com/Automattic/gp-localci-client/pull/13

I've reverted the testing commits.

**Comment by jsnajdr** [comment]
> Unlike translations, we currently have no way of telling everything using moment that it needs to re-render.

The `localize` HoC in `i18n-calypso` already has support for being reactive: it rerenders all components using it when the `i18n` object fires the `change` object.

Could we somehow integrate the `moment` locale switching into the `i18n` library? That means loading the moment locales in Calypso, the same way how we load translations now, and then pass them as a new parameter to the `i18n.setLocale()` function?

That would make `moment` managed by `i18n-calypso` and a direct `import from 'moment'` would be prohibited. Is that an option?

**Comment by sirreal** [comment]
@jsnajdr I wanted to bring your attention to https://github.com/Automattic/i18n-calypso/pull/65, which proposes removing moment from i18n-calypso completely.

**Comment by jsnajdr** [comment]
> I wanted to bring your attention to Automattic/i18n-calypso#65, which proposes removing moment from i18n-calypso completely.

I guess I don't understand the motivation to do so. @blowery says in https://github.com/Automattic/i18n-calypso/pull/65#issuecomment-405957630:

> I think bundling moment actively hurts clients of i18n-calypso.

It looks convenient to have everything with a locale bundled inside a `i18n` module and be notified when the locale changes and the locales finish async-loading. What is the part that hurts?

**Comment by blowery** [comment]
> That would make moment managed by i18n-calypso and a direct import from 'moment' would be prohibited. Is that an option?

That's the opposite of where this needs to go. We don't want to use moment as a transitive dependency.

> What is the part that hurts?

The dependency graph is what kills consumers. 

If I want to translate something, I import `i18n-calypso`, which drags in moment and moment-timezone, even if I don't want to use them. Yes, we can lazy load the locales as the ultimate consumer, as shown here, but we can't lazy load or defer the timezone load to chunks that actually need timezones. We're forcing calypso to load all of that up front when many bits don't need to do any date work.

Could we fix that with es6 modules and tree-shaking? Partially, maybe. But there's no reason to force moment on consumers of i18n-calypso. We don't provide any additional translations for moment, we're just bundling what moment already has.

**Comment by blowery** [comment]
We _used_ to use some custom translations for moment, but [that was removed](https://github.com/Automattic/i18n-calypso/commit/3bd8440796cc1c0a434a2e88b9d49a97c32b2baa). We can probably drop those strings from the translation file and should also update the docs.

**Comment by jsnajdr** [comment]
> That's the opposite of where this needs to go. We don't want to use moment as a transitive dependency.

Then, after `moment` is removed from `i18n-calypso` and the `localize` HoC no longer passes `moment` as prop to the wrapped components, how will React components use `moment` and rerender themselves on locale update?

I agree that for non-React code, wrapping `moment` inside `i18n-calypso` is not needed. The import
```js
import { moment } from 'i18n-calypso';
```
doesn't provide any added value over
```js
import moment from 'moment';
```

But for React, I think the HoC wrapper is very nice, as it can force updates on locale load:
```jsx
localize( ( { moment, translate } ) => <b>I always render with an up-to-date locale!</b> );
```

Will we have a HoC that when used, will mean: I really intend to use moment, please make it available in my chunk?
```jsx
withMoment( ( { moment } ) => <b>I also rerender when needed, and actually use moment</b> );
```

**Comment by blowery** [comment]
> Will we have a HoC that when used, will mean: I really intend to use moment, please make it available in my chunk?

We could, certainly. Ultimately I'd like to wrap the things we're doing with moment translations in components and hide the fact that we're even using moment. So things like `<FormattedDate date={ aDate } format="YYYY mm dd" />` where internally it uses something like `withMoment`.

**Comment by flootr** [comment]
resolved conflicts

**Comment by alisterscott** [comment]
The wp-desktop build fails for this PR:

```
ERROR in ./calypso/client/state/i18n/middleware.js
Module build failed: SyntaxError: /Users/distiller/wp-desktop/calypso/client/state/i18n/middleware.js: Support for the experimental syntax 'dynamicImport' isn't currently enabled (14:3):

  12 | 	if ( action && action.type === LOCALE_SET && action.localeSlug !== 'en' ) {
  13 | 		// load moment and update the locale
> 14 | 		import( 'moment' ).then( moment => {
     | 		^
  15 | 			import( /* webpackInclude: /\.js$/ */
  16 | 			`moment/locale/${ action.localeSlug }` ).then( () => {
  17 | 				moment.locale( action.localeSlug );
```

File: client/state/i18n/middleware.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Internal dependencies
+ */
+
+import { LOCALE_SET } from 'state/action-types';
+
+function updateMoment( action ) {
+	if ( action && action.type === LOCALE_SET && action.localeSlug !== 'en' ) {
+		// load moment and update the locale
+		import( 'moment' ).then( moment => {
+			import( /* webpackInclude: /\.js$/ */
```
Comment (jsnajdr) [inline client/state/i18n/middleware.js]: Nice trick! The `webpackInclude` option tells webpack to bundle all `*.js` files in the directory?

Can we also instruct Webpack to give these chunks reasonable names? It's all numbers now, `moment-locale-cs` would be a better name.

**Comment by blowery** [comment]
> The wp-desktop build fails for this PR

Likely need to update the babel config over in desktop.

**Review by jsnajdr — COMMENTED** [review]
Some issues found while playing with this and the [PR that removes Moment from i18n-calypso](https://github.com/Automattic/i18n-calypso/pull/65):

With the new version of `i18n-calypso`, `moment-timezone` is no longer loaded and `moment.tz` is undefined. Calypso uses it at a few places and will crash.

After setting the global locale with `moment.locale( 'cs' )`, existing `moment` instances will keep the previous locale. That can lead to broken UI if we store `moment` instances in Redux store or anywhere else. We should audit our `moment` usages and limit them to a few specific use cases: formatting dates, comparing dates, anything else?

There is currently no mechanism to reactively update existing UI when locale finishes loading. We should fix that before merging. Otherwise we're breaking the date/time l10n, aren't we?

**Comment by adlk** [comment]
@blowery @alisterscott this was an issue with the desktop test itself. It is fixed in https://github.com/Automattic/wp-desktop/pull/482

**Comment by blowery** [comment]
> With the new version of i18n-calypso, moment-timezone is no longer loaded and moment.tz is undefined. Calypso uses it at a few places and will crash.

I think we'd land this before landing the new i18n-calypso with moment and moment-timezone removed. They're related changes but one is not necessary for the other.

> After setting the global locale with moment.locale( 'cs' ), existing moment instances will keep the previous locale. That can lead to broken UI if we store moment instances in Redux store or anywhere else. We should audit our moment usages and limit them to a few specific use cases: formatting dates, comparing dates, anything else?

> There is currently no mechanism to reactively update existing UI when locale finishes loading. We should fix that before merging. Otherwise we're breaking the date/time l10n, aren't we?

Yeah... We should. Maybe we do that work in a separate series of PRs and then land this last? Otherwise this is going to get huuuuuuuuuge.

**Comment by blowery** [comment]
Closing this in favor of #26476

**Pull request closed without merging**